File: app/src/main/java/com/mapmyindia/sdk/demo/java/activity/PoiAlongRouteActivity.java
package com.mapmyindia.sdk.demo.java.activity;

import android.content.DialogInterface;
import android.os.Bundle;
import android.text.InputFilter;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.mapmyindia.sdk.demo.R;
import com.mapmyindia.sdk.demo.java.adapter.PoiAlongAdapter;
import com.mapmyindia.sdk.demo.java.plugin.DirectionPolylinePlugin;
import com.mapmyindia.sdk.demo.java.utils.CheckInternet;
import com.mapmyindia.sdk.demo.java.utils.InputFilterMinMax;
import com.mapmyindia.sdk.demo.java.utils.TransparentProgressDialog;
import com.mapmyindia.sdk.geojson.Point;
import com.mapmyindia.sdk.geojson.utils.PolylineUtils;
import com.mapmyindia.sdk.maps.MapView;
import com.mapmyindia.sdk.maps.MapmyIndiaMap;
import com.mapmyindia.sdk.maps.OnMapReadyCallback;
import com.mapmyindia.sdk.maps.annotations.MarkerOptions;
import com.mapmyindia.sdk.maps.camera.CameraUpdateFactory;
import com.mapmyindia.sdk.maps.geometry.LatLng;
import com.mapmyindia.sdk.maps.geometry.LatLngBounds;
import com.mmi.services.api.OnResponseCallback;
import com.mmi.services.api.alongroute.MapmyIndiaPOIAlongRoute;
import com.mmi.services.api.alongroute.MapmyIndiaPOIAlongRouteManager;
import com.mmi.services.api.alongroute.models.POIAlongRouteResponse;
import com.mmi.services.api.alongroute.models.SuggestedPOI;
import com.mmi.services.api.directions.DirectionsCriteria;
import com.mmi.services.api.directions.MapmyIndiaDirectionManager;
import com.mmi.services.api.directions.MapmyIndiaDirections;
import com.mmi.services.api.directions.models.DirectionsResponse;
import com.mmi.services.api.directions.models.DirectionsRoute;
import com.mmi.services.utils.Constants;

import java.util.ArrayList;
import java.util.List;

public class PoiAlongRouteActivity extends AppCompatActivity implements OnMapReadyCallback, MapmyIndiaMap.OnMapLongClickListener {

    private MapmyIndiaMap mapmyIndiaMap;
    private MapView mapView;
    private RecyclerView poiRecyclerView;
    private LinearLayoutManager mLayoutManager;
    private TransparentProgressDialog transparentProgressDialog;
    private String profile = DirectionsCriteria.PROFILE_DRIVING;
    private DirectionPolylinePlugin directionPolylinePlugin;
    private RelativeLayout view;
    private EditText etStartLat, etStartLng, etDestLat, etDestLng, etKeyword;
    private BottomSheetBehavior bottomSheetBehavior;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_poi_along_route);
        mapView = findViewById(R.id.map_view);
        mapView.onCreate(savedInstanceState);
        mapView.getMapAsync(this);
        initReferences();
        etStartLat = findViewById(R.id.et_start_lat);
        etStartLng = findViewById(R.id.et_start_lng);
        etDestLat = findViewById(R.id.et_dest_lat);
        etDestLng = findViewById(R.id.et_dest_lng);
        etKeyword = findViewById(R.id.et_keyword);
        etStartLng.setFilters(new InputFilter[]{new InputFilterMinMax(-180, 180)});
        etStartLat.setFilters(new InputFilter[]{new InputFilterMinMax(-90, 90)});
        etDestLng.setFilters(new InputFilter[]{new InputFilterMinMax(-180, 180)});
        etDestLat.setFilters(new InputFilter[]{new InputFilterMinMax(-90, 90)});

        view = findViewById(R.id.bottom_sheet);
        bottomSheetBehavior = BottomSheetBehavior.from(view);
        bottomSheetBehavior.setHideable(false);
        bottomSheetBehavior.setPeekHeight(250);
        bottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);

        Button button = findViewById(R.id.btn_search);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getDirections();
            }
        });
        transparentProgressDialog = new TransparentProgressDialog(this, R.drawable.circle_loader, "");

    }


    private void initReferences() {
        mLayoutManager = new LinearLayoutManager(this);
        poiRecyclerView = findViewById(R.id.poiRecyclerview);
        poiRecyclerView.setLayoutManager(mLayoutManager);
//        poiRecyclerView.setVisibility(View.GONE);

    }

    @Override
    public void onMapReady(MapmyIndiaMap mapmyIndiaMap) {
        this.mapmyIndiaMap = mapmyIndiaMap;

        mapmyIndiaMap.moveCamera(CameraUpdateFactory.newLatLngZoom(new LatLng(
                28.594475, 77.202432), 10));
        findViewById(R.id.details_layout).setVisibility(View.VISIBLE);
        mapmyIndiaMap.addOnMapLongClickListener(this);
        if (CheckInternet.isNetworkAvailable(PoiAlongRouteActivity.this)) {
            Log.v("route", "calling");
            getDirections();
        } else {
            Toast.makeText(this, getString(R.string.pleaseCheckInternetConnection), Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onMapError(int i, String s) {
    }


    private void getDirections() {

        double startLat = 0.0;
        double startLng = 0.0;
        double destLat = 0.0;
        double destLng = 0.0;
        String startLatText = etStartLat.getText().toString();
        String startLngText = etStartLng.getText().toString();
        String destLatText = etDestLat.getText().toString();
        String destLngText = etDestLng.getText().toString();
        String keyword = etKeyword.getText().toString();
        if (TextUtils.isEmpty(startLatText) || TextUtils.isEmpty(startLngText) || TextUtils.isEmpty(destLatText) || TextUtils.isEmpty(destLngText) || TextUtils.isEmpty(keyword)) {
            Toast.makeText(this, "Please fill all fields", Toast.LENGTH_SHORT).show();
            return;
        }
        try {
            startLat = Double.parseDouble(startLatText);
            startLng = Double.parseDouble(startLngText);
            destLng = Double.parseDouble(destLngText);
            destLat = Double.parseDouble(destLatText);
        } catch (Exception e) {
            //Igonore
        }
        if (startLat == 0 || startLng == 0 || destLat == 0 || destLng == 0) {
            Toast.makeText(this, "Invalid Coordinates", Toast.LENGTH_SHORT).show();
            return;
        }
        view.setVisibility(View.GONE);
        progressDialogShow();

        MapmyIndiaDirections directions = MapmyIndiaDirections.builder()
                .origin(Point.fromLngLat(startLng, startLat))
                .destination(Point.fromLngLat(destLng, destLat))
                .profile(profile)
                .resource(DirectionsCriteria.RESOURCE_ROUTE)
                .steps(false)
                .alternatives(false)
                .overview(DirectionsCriteria.OVERVIEW_FULL).build();
        MapmyIndiaDirectionManager.newInstance(directions).call(new OnResponseCallback<DirectionsResponse>() {
            @Override
            public void onSuccess(DirectionsResponse directionsResponse) {
                if (directionsResponse != null) {
                    List<DirectionsRoute> results = directionsResponse.routes();

                    if (results.size() > 0) {
                        mapmyIndiaMap.clear();
                        DirectionsRoute directionsRoute = results.get(0);
                        if (directionsRoute != null && directionsRoute.geometry() != null) {

                            drawPath(PolylineUtils.decode(directionsRoute.geometry(), Constants.PRECISION_6));

                            callPOIAlongRoute(directionsRoute.geometry(), keyword);
                        }
                    }
                }
            }

            @Override
            public void onError(int i, String s) {
                progressDialogHide();
                Toast.makeText(PoiAlongRouteActivity.this, s + "----" + i, Toast.LENGTH_LONG).show();
            }
        });
    }


    private void callPOIAlongRoute(String path, String keyword) {
        if (keyword == null) {

            return;
        }
        MapmyIndiaPOIAlongRoute poiAlongRoute = MapmyIndiaPOIAlongRoute.builder()
                .category(keyword)
                .path(path)
                .buffer(300)
                .build();
        MapmyIndiaPOIAlongRouteManager.newInstance(poiAlongRoute).call(new OnResponseCallback<POIAlongRouteResponse>() {
            @Override
            public void onSuccess(POIAlongRouteResponse poiAlongRouteResponse) {
                progressDialogHide();
                if (poiAlongRouteResponse != null) {
                    List<SuggestedPOI> pois = poiAlongRouteResponse.getSuggestedPOIs();
                    view.setVisibility(View.VISIBLE);
                    addMarker(pois);

                }
            }

            @Override
            public void onError(int i, String s) {
                progressDialogHide();
                Toast.makeText(PoiAlongRouteActivity.this, s + "----" + i, Toast.LENGTH_LONG).show();
            }
        });

    }

    private void drawPath(@NonNull List<Point> waypoints) {
        if (mapmyIndiaMap == null) {
            return;
        }
        ArrayList<LatLng> listOfLatLng = new ArrayList<>();
        for (Point point : waypoints) {
            listOfLatLng.add(new LatLng(point.latitude(), point.longitude()));
        }

        if (directionPolylinePlugin == null) {
            directionPolylinePlugin = new DirectionPolylinePlugin(mapmyIndiaMap, mapView, profile);
            directionPolylinePlugin.createPolyline(listOfLatLng);
        } else {
            directionPolylinePlugin.updatePolyline(profile, listOfLatLng);

        }
//        mapmyIndiaMap.addPolyline(new PolylineOptions().addAll(listOfLatLng).color(Color.parseColor("#3bb2d0")).width(4));
        LatLngBounds latLngBounds = new LatLngBounds.Builder().includes(listOfLatLng).build();
        mapmyIndiaMap.animateCamera(CameraUpdateFactory.newLatLngBounds(latLngBounds, 10, 250, 10, 100));
    }


    private void addMarker(List<SuggestedPOI> pois) {
        for (SuggestedPOI marker : pois) {
            mapmyIndiaMap.addMarker(new MarkerOptions().position(new LatLng(marker.getLatitude(), marker.getLongitude())).title(marker.getPoi()));
        }

        poiRecyclerView.setAdapter(new PoiAlongAdapter(pois));
    }

    /**
     * Show Progress Dialog
     */
    private void progressDialogShow() {
        transparentProgressDialog.show();
    }

    /**
     * Hide Progress dialog
     */
    private void progressDialogHide() {
        transparentProgressDialog.dismiss();
    }

    @Override
    protected void onStart() {
        super.onStart();
        mapView.onStart();
    }

    @Override
    protected void onStop() {
        super.onStop();
        mapView.onStop();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mapView.onDestroy();
    }

    @Override
    protected void onPause() {
        super.onPause();
        mapView.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        mapView.onResume();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        mapView.onLowMemory();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        mapView.onSaveInstanceState(outState);
    }

    @Override
    public boolean onMapLongClick(@NonNull LatLng latLng) {
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(this);
        alertDialog.setMessage("Select Point as Source or Destination");

        alertDialog.setPositiveButton("Source", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                etStartLat.setText(String.format("%s", latLng.getLatitude()));
                etStartLng.setText(String.format("%s", latLng.getLongitude()));
                getDirections();
            }
        });
        alertDialog.setNegativeButton("Destination", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                etDestLat.setText(String.format("%s", latLng.getLatitude()));
                etDestLng.setText(String.format("%s", latLng.getLongitude()));
                getDirections();
            }
        });

        alertDialog.setCancelable(true);
        alertDialog.show();
        return false;
    }
}

File: app/src/main/java/com/mapmyindia/sdk/demo/kotlin/activity/PoiAlongRouteActivity.kt
package com.mapmyindia.sdk.demo.kotlin.activity

import android.os.Bundle
import android.text.InputFilter
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.RelativeLayout
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.mapmyindia.sdk.demo.R
import com.mapmyindia.sdk.demo.databinding.ActivityPoiAlongRouteBinding
import com.mapmyindia.sdk.demo.java.adapter.PoiAlongAdapter
import com.mapmyindia.sdk.demo.java.plugin.DirectionPolylinePlugin
import com.mapmyindia.sdk.demo.java.utils.CheckInternet
import com.mapmyindia.sdk.demo.java.utils.InputFilterMinMax
import com.mapmyindia.sdk.demo.java.utils.TransparentProgressDialog
import com.mapmyindia.sdk.geojson.Point
import com.mapmyindia.sdk.geojson.utils.PolylineUtils
import com.mapmyindia.sdk.maps.MapView
import com.mapmyindia.sdk.maps.MapmyIndiaMap
import com.mapmyindia.sdk.maps.OnMapReadyCallback
import com.mapmyindia.sdk.maps.annotations.MarkerOptions
import com.mapmyindia.sdk.maps.camera.CameraUpdateFactory
import com.mapmyindia.sdk.maps.geometry.LatLng
import com.mapmyindia.sdk.maps.geometry.LatLngBounds
import com.mmi.services.api.OnResponseCallback
import com.mmi.services.api.alongroute.MapmyIndiaPOIAlongRoute
import com.mmi.services.api.alongroute.MapmyIndiaPOIAlongRouteManager
import com.mmi.services.api.alongroute.models.POIAlongRouteResponse
import com.mmi.services.api.alongroute.models.SuggestedPOI
import com.mmi.services.api.directions.DirectionsCriteria
import com.mmi.services.api.directions.MapmyIndiaDirectionManager
import com.mmi.services.api.directions.MapmyIndiaDirections
import com.mmi.services.api.directions.models.DirectionsResponse
import com.mmi.services.utils.Constants
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import timber.log.Timber
import java.util.*

class PoiAlongRouteActivity: AppCompatActivity(), OnMapReadyCallback, MapmyIndiaMap.OnMapLongClickListener {
    private lateinit var mBinding:ActivityPoiAlongRouteBinding
    private lateinit var mapmyIndiaMap: MapmyIndiaMap
    private lateinit var mapView: MapView
    private lateinit var transparentProgressDialog: TransparentProgressDialog
    private val profile = DirectionsCriteria.PROFILE_DRIVING

    private lateinit var poiRecyclerView: RecyclerView
    private lateinit var mLayoutManager: LinearLayoutManager
    private var directionPolylinePlugin: DirectionPolylinePlugin? = null
    private lateinit var view: RelativeLayout
    private lateinit var etStartLat: EditText
    private lateinit var etStartLng: EditText
    private lateinit var etDestLat: EditText
    private lateinit var etDestLng: EditText
    private lateinit var etKeyword: EditText
    private lateinit var bottomSheetBehavior: BottomSheetBehavior<RelativeLayout>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding= DataBindingUtil.setContentView(this,R.layout.activity_poi_along_route)
        mapView = findViewById(R.id.map_view)
        mapView.onCreate(savedInstanceState)
        mapView.getMapAsync(this)
        etStartLat = findViewById(R.id.et_start_lat)
        etStartLng = findViewById(R.id.et_start_lng)
        etDestLat = findViewById(R.id.et_dest_lat)
        etDestLng = findViewById(R.id.et_dest_lng)
        etKeyword = findViewById(R.id.et_keyword)
        etStartLng.filters = arrayOf<InputFilter>(InputFilterMinMax(-180, 180))
        etStartLat.filters = arrayOf<InputFilter>(InputFilterMinMax(-90, 90))
        etDestLng.filters = arrayOf<InputFilter>(InputFilterMinMax(-180, 180))
        etDestLat.filters = arrayOf<InputFilter>(InputFilterMinMax(-90, 90))
        initReferences()
        view = findViewById(R.id.bottom_sheet)
        bottomSheetBehavior = BottomSheetBehavior.from(view)
        bottomSheetBehavior.isHideable = false
        bottomSheetBehavior.peekHeight = 250
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED
        val button: Button = findViewById(R.id.btn_search)
        button.setOnClickListener { getDirections() }
        transparentProgressDialog = TransparentProgressDialog(this, R.drawable.circle_loader, "")
    }

    private fun initReferences() {
        poiRecyclerView = findViewById(R.id.poiRecyclerview)
        mLayoutManager = LinearLayoutManager(this)
        poiRecyclerView.layoutManager = mLayoutManager
    }

    override fun onMapError(p0: Int, p1: String?) {

    }

    override fun onMapReady(p0: MapmyIndiaMap) {
        mapmyIndiaMap = p0
        mapmyIndiaMap.moveCamera(CameraUpdateFactory.newLatLngZoom(LatLng(
                28.594475, 77.202432), 10.0))
        findViewById<View>(R.id.details_layout).visibility = View.VISIBLE
        mapmyIndiaMap.uiSettings?.setLogoMargins(0, 0, 0, 250)
        if (CheckInternet.isNetworkAvailable(this@PoiAlongRouteActivity)) {
            Timber.tag("route").v("calling")
            getDirections()
        } else {
            Toast.makeText(this, getString(R.string.pleaseCheckInternetConnection), Toast.LENGTH_SHORT).show()
        }
    }


    private fun getDirections() {
        var startLat = 0.0
        var startLng = 0.0
        var destLat = 0.0
        var destLng = 0.0
        val startLatText = etStartLat.text.toString()
        val startLngText = etStartLng.text.toString()
        val destLatText = etDestLat.text.toString()
        val destLngText = etDestLng.text.toString()
        val keyword = etKeyword.text.toString()
        if (TextUtils.isEmpty(startLatText) || TextUtils.isEmpty(startLngText) || TextUtils.isEmpty(destLatText) || TextUtils.isEmpty(destLngText) || TextUtils.isEmpty(keyword)) {
            Toast.makeText(this, "Please fill all fields", Toast.LENGTH_SHORT).show()
            return
        }
        try {
            startLat = startLatText.toDouble()
            startLng = startLngText.toDouble()
            destLng = destLngText.toDouble()
            destLat = destLatText.toDouble()
        } catch (e: Exception) {
            //Igonore
        }
        if (startLat == 0.0 || startLng == 0.0 || destLat == 0.0 || destLng == 0.0) {
            Toast.makeText(this, "Invalid Coordinates", Toast.LENGTH_SHORT).show()
            return
        }
        view.visibility = View.GONE

        progressDialogShow()
        val directions = MapmyIndiaDirections.builder()
                .origin(Point.fromLngLat(startLng, startLat))
                .destination(Point.fromLngLat(destLng, destLat))
                .profile(profile)
                .resource(DirectionsCriteria.RESOURCE_ROUTE)
                .steps(false)
                .alternatives(false)
                .overview(DirectionsCriteria.OVERVIEW_FULL).build()
        MapmyIndiaDirectionManager.newInstance(directions).call(object: OnResponseCallback<DirectionsResponse>{
            override fun onSuccess(directionsResponse: DirectionsResponse?) {
                if (directionsResponse != null) {
                    val results = directionsResponse.routes()
                    if (results.size > 0) {
                        mapmyIndiaMap.clear()
                        val directionsRoute = results[0]
                        if (directionsRoute?.geometry() != null) {
                            drawPath(PolylineUtils.decode(directionsRoute.geometry()!!, Constants.PRECISION_6))

                            callPOIAlongRoute(directionsRoute.geometry()!!, keyword)
                        } else {
                            progressDialogHide()
                        }
                    } else {
                        progressDialogHide()
                    }
                } else {
                    progressDialogHide()
                }
            }

            override fun onError(p0: Int, p1: String?) {
                progressDialogHide()
                Toast.makeText(this@PoiAlongRouteActivity, p1, Toast.LENGTH_LONG).show()
            }

        })
    }


    private fun callPOIAlongRoute(path: String, keyword: String) {
        val poiAlongRoute = MapmyIndiaPOIAlongRoute.builder()
                .category(keyword)
                .path(path)
                .buffer(300)
                .build()
        MapmyIndiaPOIAlongRouteManager.newInstance(poiAlongRoute).call(object : OnResponseCallback<POIAlongRouteResponse> {
            override fun onSuccess(response: POIAlongRouteResponse?) {
                progressDialogHide()
                if (response != null) {
                    view.visibility = View.VISIBLE
                    val pois = response.suggestedPOIs
                    addMarker(pois)

                }
            }

            override fun onError(p0: Int, p1: String?) {
                progressDialogHide()
                Toast.makeText(this@PoiAlongRouteActivity, p1, Toast.LENGTH_LONG).show()
            }

        })

    }


    private fun drawPath(waypoints: List<Point>) {
        val listOfLatlang = ArrayList<LatLng>()
        for (point in waypoints) {
            listOfLatlang.add(LatLng(point.latitude(), point.longitude()))
        }

        if(directionPolylinePlugin == null) {
            directionPolylinePlugin =DirectionPolylinePlugin(mapmyIndiaMap!!, mapView!!, profile)
            directionPolylinePlugin!!.createPolyline(listOfLatlang)
        } else {
            directionPolylinePlugin!!.updatePolyline(profile, listOfLatlang)
        }

        val latLngBounds = LatLngBounds.Builder().includes(listOfLatlang).build()
        mapmyIndiaMap.animateCamera(CameraUpdateFactory.newLatLngBounds(latLngBounds, 10, 250, 10, 100))
    }


    private fun addMarker(pois: List<SuggestedPOI>) {
        for (marker in pois) {
            mapmyIndiaMap.addMarker(MarkerOptions().position(LatLng(marker.latitude, marker.longitude)).title(marker.poi))
        }
        poiRecyclerView.adapter = PoiAlongAdapter(pois)
    }

    private fun progressDialogShow() {
        transparentProgressDialog.show()
    }

    private fun progressDialogHide() {
        transparentProgressDialog.dismiss()
    }

    override fun onStart() {
        super.onStart()
        mBinding.mapView.onStart()
    }

    override fun onResume() {
        super.onResume()
        mBinding.mapView.onResume()
    }

    override fun onPause() {
        super.onPause()
        mBinding.mapView.onPause()
    }

    override fun onStop() {
        super.onStop()
        mBinding.mapView.onStop()
    }

    override fun onDestroy() {
        super.onDestroy()
        mBinding.mapView.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mBinding.mapView.onLowMemory()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        mBinding.mapView.onSaveInstanceState(outState)
    }

    override fun onMapLongClick(latLng: LatLng): Boolean {
        val alertDialog = AlertDialog.Builder(this)
        alertDialog.setMessage("Select Point as Source or Destination")

        alertDialog.setPositiveButton("Source") { dialog, which ->
            etStartLat.setText("${latLng.latitude}")
            etStartLng.setText("${latLng.longitude}")
            getDirections()
        }
        alertDialog.setNegativeButton("Destination") { dialog, which ->
            etDestLat.setText("${latLng.latitude}")
            etDestLng.setText("${latLng.longitude}")
            getDirections()
        }

        alertDialog.setCancelable(true)
        alertDialog.show()
        return false
    }
}